File: resources/evaluators/microservice/hotelcorrectness/catalog.py
"""Pinned public Hotel Reservation fixture catalog."""

# ruff: noqa: PLR2004, TRY003

from __future__ import annotations

import struct
from dataclasses import dataclass


@dataclass(frozen=True)
class Profile:
    """One externally observable seeded hotel profile."""

    id: str
    name: str
    phone: str
    lat: float
    lon: float
    recommend_lat: float
    recommend_lon: float


def _float32(value: float) -> float:
    return struct.unpack("f", struct.pack("f", value))[0]


_INITIAL_PROFILES = (
    Profile(
        "1",
        "Clift Hotel",
        "(415) 775-4700",
        _float32(37.7867),
        _float32(-122.4112),
        37.7867,
        -122.4112,
    ),
    Profile(
        "2",
        "W San Francisco",
        "(415) 777-5300",
        _float32(37.7854),
        _float32(-122.4005),
        37.7854,
        -122.4005,
    ),
    Profile(
        "3",
        "Hotel Zetta",
        "(415) 543-8555",
        _float32(37.7834),
        _float32(-122.4071),
        37.7834,
        -122.4071,
    ),
    Profile(
        "4",
        "Hotel Vitale",
        "(415) 278-3700",
        _float32(37.7936),
        _float32(-122.3930),
        37.7936,
        -122.3930,
    ),
    Profile(
        "5",
        "Phoenix Hotel",
        "(415) 776-1380",
        _float32(37.7831),
        _float32(-122.4181),
        37.7831,
        -122.4181,
    ),
    Profile(
        "6",
        "St. Regis San Francisco",
        "(415) 284-4000",
        _float32(37.7863),
        _float32(-122.4015),
        37.7863,
        -122.4015,
    ),
)


def seed_profiles() -> dict[str, Profile]:
    """Reproduce the pinned profile and recommendation service fixtures."""
    catalog = {profile.id: profile for profile in _INITIAL_PROFILES}
    for numeric_id in range(7, 81):
        hotel_id = str(numeric_id)
        fraction = _float32(_float32(numeric_id) / _float32(500.0))
        profile_lat = _float32(_float32(37.7835) + _float32(fraction * _float32(3.0)))
        profile_lon = _float32(_float32(-122.41) + _float32(fraction * _float32(4.0)))
        catalog[hotel_id] = Profile(
            hotel_id,
            "St. Regis San Francisco",
            f"(415) 284-40{hotel_id}",
            profile_lat,
            profile_lon,
            37.7835 + numeric_id / 500.0 * 3,
            -122.41 + numeric_id / 500.0 * 4,
        )
    if len(catalog) != 80:
        raise ValueError(f"hotel seed catalog has {len(catalog)} profiles, expected 80")
    return catalog


def rate_backed_ids() -> frozenset[str]:
    """Return the exact hotel IDs with seeded rate plans."""
    return frozenset({"1", "2", "3", *(str(value) for value in range(9, 81, 3))})


def capacity_for_hotel(hotel_id: str) -> int:
    """Return the pinned room capacity for one catalog hotel."""
    try:
        numeric_id = int(hotel_id)
    except ValueError as error:
        raise ValueError(f"invalid hotel ID {hotel_id!r}") from error
    if not 1 <= numeric_id <= 80:
        raise ValueError(f"hotel ID {hotel_id!r} is outside the seeded catalog")
    if numeric_id <= 6 or numeric_id % 3 == 0:
        return 200
    return 300 if numeric_id % 3 == 1 else 250

File: examples/microservices/hotel-correctness/evaluator/internal/hotel/catalog.go
package hotel

import (
	"fmt"
	"strconv"
)

type profile struct {
	id, name, phone string
	lat, lon        float32
	recommendLat    float64
	recommendLon    float64
}

var initialProfiles = []profile{
	{id: "1", name: "Clift Hotel", phone: "(415) 775-4700", lat: 37.7867, lon: -122.4112, recommendLat: 37.7867, recommendLon: -122.4112},
	{id: "2", name: "W San Francisco", phone: "(415) 777-5300", lat: 37.7854, lon: -122.4005, recommendLat: 37.7854, recommendLon: -122.4005},
	{id: "3", name: "Hotel Zetta", phone: "(415) 543-8555", lat: 37.7834, lon: -122.4071, recommendLat: 37.7834, recommendLon: -122.4071},
	{id: "4", name: "Hotel Vitale", phone: "(415) 278-3700", lat: 37.7936, lon: -122.3930, recommendLat: 37.7936, recommendLon: -122.3930},
	{id: "5", name: "Phoenix Hotel", phone: "(415) 776-1380", lat: 37.7831, lon: -122.4181, recommendLat: 37.7831, recommendLon: -122.4181},
	{id: "6", name: "St. Regis San Francisco", phone: "(415) 284-4000", lat: 37.7863, lon: -122.4015, recommendLat: 37.7863, recommendLon: -122.4015},
}

// seedProfiles independently reproduces cmd/profile/db.go. In particular, the
// profile service stores generated coordinates as float32 while the
// recommendation service stores its copy as float64.
func seedProfiles() (map[string]profile, error) {
	catalog := make(map[string]profile, 80)
	for _, item := range initialProfiles {
		catalog[item.id] = item
	}
	for id := 7; id <= 80; id++ {
		text := strconv.Itoa(id)
		item := profile{
			id: text, name: "St. Regis San Francisco",
			phone:        fmt.Sprintf("(415) 284-40%s", text),
			lat:          37.7835 + float32(id)/500.0*3,
			lon:          -122.41 + float32(id)/500.0*4,
			recommendLat: 37.7835 + float64(id)/500.0*3,
			recommendLon: -122.41 + float64(id)/500.0*4,
		}
		catalog[text] = item
	}
	if len(catalog) != 80 {
		return nil, fmt.Errorf("Hotel seed catalog has %d profiles, expected 80", len(catalog))
	}
	return catalog, nil
}

func rateBackedIDs() map[string]struct{} {
	result := map[string]struct{}{"1": {}, "2": {}, "3": {}}
	for id := 9; id <= 80; id += 3 {
		result[strconv.Itoa(id)] = struct{}{}
	}
	return result
}
